river/tree/nodes/sgt_nodes.py

```diff
@@ -0,0 +1,223 @@
+    prediction
+        The initial prediction of the leaf.
+    depth
+        The depth of the leaf in the tree.
+    split_params
+        Parameters passed to the feature quantizers.
+    """
+
+    def __init__(self, prediction=0.0, depth=0, split_params=None):
+        super().__init__()
+        self._prediction = prediction
+        self.depth = depth
+
+        # Quantizer params are dynamically updated
+        self.split_params = (
+            split_params if split_params is not None else collections.defaultdict(dict)
+        )
+        self.last_split_attempt_at = 0
+
+        self._split_stats: Optional[
+            Dict[
+                FeatureName,
+                Union[Dict[Hashable, GradHessStats], DynamicQuantizer, StaticQuantizer],
+            ]
+        ] = {}
+        self._update_stats = GradHessStats()
+
+    def reset(self):
+        self._split_stats = {}
+        self._update_stats = GradHessStats()
+
+    def update(self, x: dict, gh: GradHess, sgt, w: float = 1.0):
+        for idx, x_val in x.items():
+            if not isinstance(x_val, numbers.Number) or idx in sgt.nominal_attributes:
+                # Update the set of nominal features
+                sgt.nominal_attributes.add(idx)
+                try:
+                    self._split_stats[idx][x_val].update(gh, w)
+                except KeyError:
+                    if idx not in self._split_stats:
+                        # Categorical features are treated with a simple dict structure
+                        self._split_stats[idx] = {}
+                    self._split_stats[idx][x_val] = GradHessStats()
+                    self._split_stats[idx][x_val].update(gh, w)
+            else:
+                try:
+                    self._split_stats[idx].update(x_val, gh, w)
+                except KeyError:
+                    # Create a new quantizer
+                    self._split_stats[idx] = sgt.feature_quantizer._set_params(
+                        self.split_params[idx]
+                    )
+                    self._split_stats[idx].update(x_val, gh, w)
+
+        self._update_stats.update(gh, w=w)
+
+    def prediction(self) -> float:
+        return self._prediction
+
+    def find_best_split(self, sgt) -> BranchFactory:
+        best_split = BranchFactory()
+        best_split.merit = GradHessMerit()
+
+        # Null split: update the prediction using the new gradient information
+        best_split.merit.delta_pred = self.delta_prediction(
+            self._update_stats.mean, sgt.lambda_value
+        )
+        dlms = self._update_stats.delta_loss_mean_var(best_split.merit.delta_pred)
+        best_split.merit.loss_mean = dlms.mean.get()
+        best_split.merit.loss_var = dlms.get()
+
+        for feature_idx in self._split_stats:
+            candidate = BranchFactory()
+            candidate.merit = GradHessMerit()
+            candidate.feature = feature_idx
+
+            if feature_idx in sgt.nominal_attributes:
```

Comment:
I would break this function down into two functions: one for categorical and one for numerical. But again, no biggie :)

Reply:
Nice catch! I totally agree. The end result is much simpler to read. You can check it at [this commit](https://github.com/online-ml/river/pull/674/commits/b33ed26aa625562c6df2c6e1bda4ebb59616a72c)